### PR TITLE
Allow smart contract verification

### PR DESCRIPTION
**Pull request by shargon**
As seen in the talk with @lock9 

- Allow to see watch only address
- Allow to use smart contract as sender
- Remove unused method.

Require https://github.com/neo-project/neo/pull/1800

### CODE REVIEW

File: neo-cli/CLI/MainService.Contracts.cs [hunk 1]
```diff
@@ -53,7 +53,7 @@ private void OnInvokeCommand(UInt160 scriptHash, string operation, JArray contra
         {
             Signer[] signers = Array.Empty<Signer>();
             if (signerAccounts != null && !NoWallet())
-                signers = CurrentWallet.GetAccounts().Where(p => !p.Lock && !p.WatchOnly && signerAccounts.Contains(p.ScriptHash)).Select(p => new Signer() { Account = p.ScriptHash, Scopes = WitnessScope.CalledByEntry }).ToArray();
```
Comment (superboyiii) [inline neo-cli/CLI/MainService.Contracts.cs:56]: Why `WatchOnly` address can be signer?

File: neo-cli/CLI/MainService.Contracts.cs [hunk 2]
```diff
@@ -53,7 +48,7 @@ private void OnInvokeCommand(UInt160 scriptHash, string operation, JArray contra
...
+                signers = CurrentWallet.GetAccounts().Where(p => !p.Lock && signerAccounts.Contains(p.ScriptHash)).Select(p => new Signer() { Account = p.ScriptHash, Scopes = WitnessScope.CalledByEntry }).ToArray();
```
Comment (superboyiii) [inline neo-cli/CLI/MainService.Contracts.cs:51]: Why WatchOnly address can be signer?

Comment (shargon) [reply neo-cli/CLI/MainService.Contracts.cs:51]: Withdrawal from a contract require a watch only address.

**Comment by shargon** [comment]
@lock9 could you confirm that it solve your problem?

File: neo-cli/CLI/MainService.Wallet.cs [hunk 3]
```diff
@@ -284,6 +284,57 @@ private void OnImportKeyCommand(string wifOrFile)
...
+        [ConsoleCommand("import watchonly", Category = "Wallet Commands")]
+        private void OnImportWatchOnlyCommand(string addressOrFile)
+        {
+            UInt160 address = null;
+            try
+            {
+                address = StringToAddress(addressOrFile);
+            }
+            catch (FormatException) { }
+            if (address == null)
```
Comment (Tommo-L) [inline neo-cli/CLI/MainService.Wallet.cs:299]:
```suggestion
            if (address is null)
```

**Comment by Tommo-L** [comment]
If we want to allow to use smart contract as sender, we must modify neo-core first.

https://github.com/neo-project/neo/blob/master/src/neo/Wallets/Wallet.cs#L252

https://github.com/neo-project/neo/blob/master/src/neo/Wallets/Wallet.cs#L320

https://github.com/neo-project/neo/blob/master/src/neo/Wallets/Wallet.cs#L396

https://github.com/neo-project/neo/blob/master/src/neo/Wallets/Wallet.cs#L402

**Comment by lock9** [comment]
@shargon  I can only test this tomorrow morning, by only checking the code I can't ensure it works

**Comment by shargon** [comment]
> If we want to allow to use smart contract as sender, we must modify neo-core first.

You should add as your smart contract as watch only first.

**Comment by ProDog** [comment]
I have tested and it throw an exception, if I `import watchonly` a smart contract hash, because it's key is null.
```
neo> import watchonly 0xb6770c896f96f3d85381ec45a403f37d18f6d714
Address: NMpBQr7KMPoVd82LeJbVYnH35m8jxWvhZh
```

image

**Comment by shargon** [comment]
Yes, we can't know the pubkey

**Comment by meevee98** [comment]
I've tried to withdraw gas from a smart contract by using the transfer command.
The validation is working well, until the moment the cli asks for relaying the transaction. If I relay, it fails to send the transaction.

image

As @Tommo-L [has said](https://github.com/neo-project/neo-node/pull/628#issuecomment-663335254), the neo-core must be changed as well. I've been researching and this fail is because the ``Wallet.Sign``, that only has validation for multi signature or accounts with public key. Including a smart contract as a watch-only address doesn't include a public key, so a transaction signing fails.

https://github.com/neo-project/neo/blob/daa6e7a3b95524534107cfb6e174d9a52f3cd617/src/neo/Wallets/Wallet.cs#L414-L440

**Comment by shargon** [comment]
@meevee98 @Tommo-L please, could you check https://github.com/neo-project/neo/pull/1800 ?

**Comment by ProDog** [comment]
I have tested it OK.
But I have a question, why must `import watchonly` first if add contract as verification? For developers, they would like to make it easier to use this, such as using the SDK to construct this transaction.

**Comment by shargon** [comment]
@ProDog now with https://github.com/neo-project/neo/pull/1800#discussion_r463935448 (https://github.com/neo-project/neo/pull/1800/commits/ab971cede24d08dca66986ecc201ca7c0208023d) it's not required to add the contract as `watch-only`.

**Comment by ProDog** [comment]
@erikzhang this need to merge, please review it at sometime.

**Comment by erikzhang** [comment]
Compilation error.

**Comment by shargon** [comment]
merge?

**Pull request merged**